### DISCUSSION TITLE
Diagnostic errors are not shown when integrating with Webpack

### DISCUSSION THREAD

**Discussion by Isqanderm** [nonara/ts-patch#146] [Q&A] [answered]
Hi, I'm trying to make my transformer.

Expected behavior: I was waiting for typing errors or other diagnostic messages to be shown during the webpack build process, since transformer works fine when compiled via tsc.

Question: Tell me, maybe this is an unsolvable problem or the problem needs to be solved in some other way?  I will be grateful for the direction.

Example of errors: 
<img width="1205" alt="Снимок экрана 2024-02-15 в 23 12 31" src="https://github.com/nonara/ts-patch/assets/152389998/426ab09c-9245-40f9-a254-48ffe648c89c">

I made a small example to reproduce: https://github.com/Isqanderm/onion-transformer/tree/main/examples/webpack

P.S.: transformer itself works perfectly. This approach opens up vast horizons for imagination. Thanks for that.

**Comment by nonara** [comment]
Hi Aleksandr!

If you're using `ts-loader`, you should be able to set the compiler to `ts-patch/compiler`

If you use another loader, check it if has a compiler option (it should), and you can point it to tsp.

Hope that helps!

**Reply by Isqanderm** [reply]
I'm definitely using ts-patch/compiler along with ts-loader. it looks more like ts-patch/compiler does not show custom errors when integrating with webpack, because the errors of the TS code are usually perfectly visible. It feels like ts-patch/compiler or ts-loader itself is "eating up" my error messages.

<img width="838" alt="Снимок экрана 2024-02-16 в 20 48 41" src="https://github.com/nonara/ts-patch/assets/152389998/7ad7c48d-5e52-46b5-88be-0f2727da3ef5">

**Reply by nonara** [reply, marked as answer]
Ah, sorry, I misunderstood. I missed that you were using the `addDiagnostic` feature to add custom errors.

Unfortunately, that's a part of a known issue. While the feature works correctly, it was built to work with the standard `emitFilesAndReportErrors` (iirc) function which tsc and most others that use the compiler API use for emit. 

Some libraries do some custom magic and use other means, which means our diagnostics utils aren't able to inject into the arrays produced by those methods.

The upcoming new major version of tsp will have a a more robust and higher insertion point to ensure that we can tap into diagnostics no matter where they're created.

In the mean time, it may be possible to workaround by using a Program transformer, where you take the created Program and call the getDiagnostics (or an emit method without write) first to ensure that those get generated first. I believe they're cached, so you could potentially directly modify the arrays. If memory serves, the getDiagnostics methods don't return array copies (but I'm not sure on that).

I think there are some other issues with discussion surrounding this area, which may be of some help.

Hopefully I'll have the new TSP fairly soon! Making some shifts in my work schedule so I can focus more on OSS.
